HLTrigger/special/plugins/HLTLogMonitorFilter.cc

```diff
@@ -33,74 +34,95 @@
 // class declaration
 //
 
-class HLTLogMonitorFilter : public edm::EDFilter {
+class HLTLogMonitorFilter : public edm::global::EDFilter<> {
 public:
   explicit HLTLogMonitorFilter(const edm::ParameterSet&);
   ~HLTLogMonitorFilter() override;
 
   struct CategoryEntry {
     uint32_t
         threshold;  // configurable threshold, after which messages in this Category start to be logarithmically prescaled
-    uint32_t prescale;  // current prescale for this Category
-    uint64_t counter;   // number of events that fired this Category
-    uint64_t accepted;  // number of events acepted for this Category
-    bool done;          // track if this Category has already been seen in the current event
+    std::atomic<uint32_t> maxPrescale;  // maximum prescale used for this Category
+    std::atomic<uint64_t> counter;      // number of events that fired this Category
+    std::atomic<uint64_t> accepted;     // number of events acepted for this Category
+    uint32_t id;                        // monotonically increasing unique id for this Category
 
-    CategoryEntry(uint32_t t = 0)
+    CategoryEntry(uint32_t iID = 0, uint32_t t = 0)
         : threshold(
               t),  // default-constructed entries have the threshold set to 0, which means the associated Category is disabled
-          prescale(1),
+          maxPrescale(1),
           counter(0),
           accepted(0),
-          done(false) {}
+          id(iID) {}
 
+    CategoryEntry(CategoryEntry const& iOther)
+        : threshold(iOther.threshold),
+          maxPrescale(iOther.maxPrescale.load()),
+          counter(iOther.counter.load()),
+          accepted(iOther.counter.load()),
+          id(iOther.id) {}
+
+    // caller guarantees this is called just once per event
     bool accept() {
-      if (not done) {
-        done = true;
-        ++counter;
-      }
+      auto tCounter = ++counter;
 
       // fail if the prescaler is disabled (threshold == 0), or if the counter is not a multiple of the prescale
-      if ((threshold == 0) or (counter % prescale))
+      if (threshold == 0) {
         return false;
+      }
+
+      if (threshold == 1) {
+        ++accepted;
+        return true;
+      }
 
-      // quasi-logarithmic increase in the prescale factor (should be safe even if threshold is 1)
-      if (counter == prescale * threshold)
-        prescale *= threshold;
+      uint64_t dynPrescale = 1;
+      // quasi-logarithmic increase in the prescale factor
+      // dynamically calculating the prescale is mulit-thread stable
+      while (tCounter > dynPrescale * threshold) {
+        dynPrescale *= threshold;
+      }
+
+      auto tMaxPrescale = maxPrescale.load();
+      while (tMaxPrescale < dynPrescale) {
+        maxPrescale.compare_exchange_strong(tMaxPrescale, dynPrescale);
+      }
```

Comment:
```suggestion
      while (maxPrescale < dynPrescale) {
        auto tMaxPrescale = maxPrescale.load();
        maxPrescale.compare_exchange_strong(tMaxPrescale, dynPrescale);
      }
```
Not really a suggestion; I was just naively expecting the green version instead. What am I missing?

Reply:
Got it. Thanks!